doc/mkfs.reiser4.8
.\"						Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH mkfs.reiser4 8 "02 Oct, 2002" reiser4progs "reiser4progs manual"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
mkfs.reiser4 \- the program for creating reiser4 filesystem.
.SH SYNOPSIS
.B mkfs.reiser4
[ options ] FILE1 FILE2 ... [ size[K|M|G] ]
.SH DESCRIPTION
.B mkfs.reiser4
is reiser4 filesystem creation program. It is based on new libreiser4
library. Since libreiser4 is fully plugin-based, we have the potential
to create not just reiser4 partitions, but any filesystem or database
format, which is based on balanced trees.
.SH COMMON OPTIONS
.TP
.B -V, --version
prints program version.
.TP
.B -?, -h, --help
prints program help.
.TP
.B -y, --yes
assumes an answer 'yes' to all questions.
.TP
.B -f, --force
forces mkfs to use whole disk, not block device or mounted partition.
.SH MKFS OPTIONS
.TP
.B -b, --block-size N
block size to be used (architecture page size by default)
.TP
.B -L, --label LABEL
volume label to be used
.TP
.B -U, --uuid UUID
universally unique identifier to be used
.TP
.B -s, --lost-found
forces mkfs to create lost+found directory.
.TP
.B -d, --discard
tells mkfs to discard given device before creating the filesystem (for
solid state drives).
.SH PLUGIN OPTIONS
.TP
.B -p, --print-profile
prints the plugin profile. This is the set of default plugins used for all
parts of a filesystem -- format, nodes, files, directories, hashes, etc. If
--override is specified, then prints modified plugins.
.TP
.B -l, --print-plugins
prints all plugins libreiser4 know about.
.TP
.B -o, --override TYPE=PLUGIN, ...
overrides the default plugin of the type "TYPE" by the plugin "PLUGIN"
in the plugin profile.
.sp 1
Examples:
.sp 1
assign short key plugin to "key" field in order to create filesystem
with short keys policy:
.sp 1
mkfs.reiser4 -yf -o key=key_short /dev/hda2
.RS
.SH REPORTING BUGS
Report bugs to <reiserfs-devel@vger.kernel.org>
.SH SEE ALSO
.BR measurefs.reiser4(8),
.BR debugfs.reiser4(8),
.BR fsck.reiser4(8)
.SH AUTHOR
This manual page was written by Yury Umanets <umka@namesys.com>



ChangeLog
This file is part of reiser4progs package

Copyright (C) 2001, 2002, 2003, 2004 by Hans Reiser, licensing governed by
reiser4progs/COPYING.

This file may be modified and/or distributed without restriction. This is
not an invitation to misrepresent the history of reiser4progs.

2015-08-14
	1.1.0 version
2015-08-14
	Format release 4.0.1;
	Support checksum for formatted nodes;
	NODE41 format with 32-bit field for checksums;
2014-06-28
	1.0.9 version
2014-06-28
	discard option to mkfs -- discard device at filesystem creation
	time (for SSD drives)

	Fixup in librepair/twig_scan.c: cb_item_region_check().
2013-05-03
	1.0.8 version
2013-05-03
	Fixups in fsck.reiser4: repair a special kind of corruptions
	appeared due to software bugs in earlier kernels.

	mkfs.reiser4 now enables intelligent compression with lzo1
	algorithm by default. If you want to create reiser4 without
	compression, then override the default plugin set with the
	"-o create=reg40" option.
2008-01-09
	1.0.7 version.
2008-01-09
	Alias -p "preen" to -a "auto" for fsck.
	Bugfixes:
	- fixed bugs in CryptoCompression object plugin (incorrect handling
	of corrupted compressed bodies).
	- suppress messages about wrong bytes for CryptoCompression
	objects	as in most cases it is because kernel doesn't support
	i_blocks and i_bytes for such objects because of performance
	issues. If wrong bytes is the result of fatal corruption, then
	fsck will detect and report it.
2006-02-22
	1.0.6 version.
2006-01-13
	Version control data is added into format40 disk format plugin.
	Version control data is added into backup blocks.
	Recover filesystems of both old and new formats - temporary.
	UPDATE_BACKUP_IS_NEEDED is added -- handle correctly cases when 
	disk format is updated and backup blocks are not; fsck correctly.
2005-12-09
	Heir plugin SET (HSET) is added.
2005-11-21
	Bugfixes:
	- a lot of fixes in compression tail item plugin;
	- a lot of fixes in CryptoCompression object plugin;
	- a lot of fixes in common object plugin repair code;
	- in spec file.
2005-11-08
	A lot of speedups.
	A lot of bugfixes.
	Demos/busy program update: crypto compression test is added.
2005-10-31
	Bugfixes in truncate, clubber and check_struct methods of all object
	plugins.
2005-10-28
	Compression plugin is added. 
	CryptoCompression plugin is adjusted to handle compressed files.
2005-10-14
	Bugfixes:
	- check the path given on a symlink creation;
	- check the possible file position overflow;
	- handle the cluster size of ctail items correctly.
2005-09-31
	Crypto-Compression Object plugins is added.
	Fix the object creation: inherit always from the parent.
	Fix the pset storage policy.
2005-08-31
	Compression tail item (ctail) plugin is added.
2005-08-16
	The full support for Crypto-Compression plugins is added into PluginSet
	and libreiser4 profile.
	Crypto SD extention is edded.
2005-08-12
	1.0.5 version.
2005-08-10
	Installation fix.
	Return value fixes.
	Backup block handling fixes.
2005-06-28
	Ability to link against libaal placed at arbitrary location is fixed.
	Unlink dir fix.
	Stat40 check_struct fix.
	Man pages for not ready utils are removed.
	Handle new (crc related) fields in plugin set correctly.
	Flush fs metadata fix (interrupt related).
	Do not flush dirty nodes if fsck fails.
	Recalculate bitmap checksums at the end of fsck if they are wrong.
	Do not update bitmaps after every fsck pass if no metadata found.
2005-03-23
	nomkid option to fsck -- ability to recover reiser4 w/out mkid check.
2005-03-21
	Large file handling fixes.
2005-03-16
	Some casting fixes.
2005-02-21
	Reg40 file body convertion fixes.
2005-02-20
	The 1.0.4 version.
2005-02-19
	A bugfix in file body convertion code.
2005-02-17
	Enable libminimal by default.
	Do not configure empty utils(resizer, cpfs).
2005-01-30
	Some bug fixes in tree balancing, fs check, syncing code.
2005-01-25
	Some bug fixes in tree balancing code.
2005-01-20
	Bug fixes:
	- in backup layout traverse code;
	- in insert extent item-by-item code;
2005-01-17
	A way to recover fs metadata independently of the mkfs id is added.
	Bug fixes:
	- in repair code then a node has a pointer to itself;
	- in not aligned access to different object on-disk data.
2004-12-23
	Bug fixes:
	- in initializing object id fields in SB;
	- in free blocks accounting;
	- in journal unpack code.
2004-12-12
	Some bugfixes:
	- in object/extent truncate code are fixed.
	- in backup handling code is fixed.
	- sware about libaal version correctly.
2004-12-04
	The version is 1.0.3
2004-12-01
	The Pack/unpack journal code is written.
	Cleanups in program options.
2004-11-25
	Build either all or none progs statically.
	The recover from unused space code is added. Could be used as undelete.
2004-11-24
	The backup code is written.
2004-10-29
	A bunch of bugs is fixed.
2004-10-21
	Object detach/unlink methods are fixed.
	Fix the handling of special entries ('.', '..') in directory objects.
	The version is 1.0.2.
2004-10-15
	File body write code (namely extent40 write) is rewritten
	and does work now.
2004-10-12
	Make allocator CRC endian safe
2004-10-07
	A bunch of bugs is fixed in the balancing code
2004-09-22
	Change "stand alone" to "minimal footprint"
2004-09-21
	Get rid of key collision problem is tree_scan
2004-09-16
	sd_bytes fix for files larger then 4G.
2004-09-14
	Make ldconfig to run automatically on 'make install'.
2004-09-13
	Make r5 hash working correctly with extended ASCII characters.
2004-08-24
	Some bugs are fixed:
	- in tree_scan method;
	- in reading extent items;
	- in man pages.
	Some optimisations for stand-alone are added.
	Version is 1.0.1.
2004-08-09
	1.0.0 version is released.


progs/mkfs/mkfs.c
/* Copyright (C) 2001-2005 by Hans Reiser, licensing governed by
   reiser4progs/COPYING.
   
   mkfs.c -- program for creating reiser4 filesystem. */

#ifdef HAVE_CONFIG_H
#  include <config.h> 
#endif

#if defined(HAVE_LIBUUID) && defined(HAVE_UUID_UUID_H)
#  include <uuid/uuid.h>
#endif

#ifndef ENABLE_MINIMAL
#  include <time.h>
#  include <stdlib.h>
#endif

#include <stdio.h>
#include <string.h>
#include <fcntl.h>
#include <getopt.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>
#include <sys/stat.h>

#ifdef HAVE_UNAME
#  include <sys/utsname.h>
#endif

#include <reiser4/libreiser4.h>

#include <aux/aux.h>
#include <misc/misc.h>

typedef enum mkfs_behav_flags {
	BF_FORCE      = 1 << 0,
	BF_YES        = 1 << 1,
	BF_LOST       = 1 << 2,
	BF_SHOW_PARM  = 1 << 3,
	BF_SHOW_PLUG  = 1 << 4,
	BF_DISCARD    = 1 << 5,
	BF_MIRRORS    = 1 << 6
} mkfs_behav_flags_t;

/* Prints mkfs options */
static void mkfs_print_usage(char *name) {
	fprintf(stderr, "Usage: %s [ options ] "
		"FILE1 FILE2 ... [ size[K|M|G] ]\n", name);
    
	fprintf(stderr, 
		"Mkfs options:\n"
		"  -s, --lost-found              forces mkfs to create lost+found\n"
		"                                directory.\n"
		"  -b, --block-size N            block size, 4096 by default, other\n"
		"                                are not supported at the moment.\n"
		"  -U, --uuid UUID               universally unique identifier.\n"
		"  -L, --label LABEL             volume label lets to mount\n"
		"                                filesystem by its label.\n"
		"Plugins options:\n"
		"  -p, --print-profile           prints the plugin profile.\n"
		"  -l, --print-plugins           prints all known plugins.\n"
	        "  -o, --override TYPE=PLUGIN    overrides the default plugin of the type\n"
	        "                                \"TYPE\" by the plugin \"PLUGIN\" in the\n"
		"                                profile.\n"
		"Common options:\n"
		"  -?, -h, --help                prints program usage.\n"
		"  -V, --version                 prints current version.\n"
		"  -y, --yes                     assumes an answer 'yes' to all questions.\n"
		"  -f, --force                   makes mkfs to use whole disk, not\n"
		"                                block device or mounted partition.\n"
		"  -d, --discard                 tells mkfs to discard given device\n"
		"                                before creating the filesystem (for SSDs)\n"
		"  -m, --mirrors                 create mirrors on listed devices.\n");
}

/* Initializes used by mkfs exception streams */
static void mkfs_init(void) {
	int ex;

	/* Setting up exception streams*/
	for (ex = 0; ex < EXCEPTION_TYPE_LAST; ex++)
		misc_exception_set_stream(ex, stderr);
}

static reiser4_object_t *reiser4_root_create(reiser4_fs_t *fs) {
	entry_hint_t entry;
	object_info_t info;
	
	aal_assert("vpf-1625", fs != NULL);
	aal_assert("vpf-1626", fs->tree != NULL);
	
	aal_memset(&info, 0, sizeof(info));

	info.tree = (tree_entity_t *)fs->tree;

	/* Preparing entry hint. */
	entry.name[0] = '\0';
	aal_memcpy(&entry.offset, &fs->tree->key, sizeof(entry.offset));

	reiser4_pset_root(&info);
	
	return reiser4_object_create(&entry, &info, NULL);
}

int main(int argc, char *argv[]) {
	int c;
	struct stat st;

	fs_hint_t hint;
	reiser4_fs_t *fs;
	char override[4096];
	aal_device_t *device;

	aal_list_t *walk = NULL;
	aal_gauge_t *gauge = NULL;
	aal_list_t *devices = NULL;

	char *host_dev;
	count_t dev_len = 0;
	uint32_t dev_cnt = 0;
	
#ifdef HAVE_UNAME
	struct utsname sysinfo;
#endif
	
	mkfs_behav_flags_t flags = 0;
    
	static struct option long_options[] = {
		{"version", no_argument, NULL, 'V'},
		{"help", no_argument, NULL, 'h'},
		{"force", no_argument, NULL, 'f'},
		{"yes", no_argument, NULL, 'y'},
		{"block-size", required_argument, NULL, 'b'},
		{"label", required_argument, NULL, 'L'},
		{"uuid", required_argument, NULL, 'U'},
		{"lost-found", required_argument, NULL, 's'},
		{"print-profile", no_argument, NULL, 'p'},
		{"print-plugins", no_argument, NULL, 'l'},
		{"override", required_argument, NULL, 'o'},
		{"discard", no_argument, NULL, 'd'},
		{"mirrors", no_argument, NULL, 'm'},
		{0, 0, 0, 0}
	};
    
	mkfs_init();

	if (argc < 2) {
		mkfs_print_usage(argv[0]);
		return USER_ERROR;
	}

	memset(&hint, 0, sizeof(hint));
	memset(override, 0, sizeof(override));

	/* Parsing parameters */    
	while ((c = getopt_long(argc, argv, "hVyfb:U:L:splo:dm?",
				long_options, (int *)0)) != EOF) 
	{
		switch (c) {
		case 'h':
		case '?':
			mkfs_print_usage(argv[0]);
			return NO_ERROR;
		case 'V':
			misc_print_banner(argv[0]);
			return NO_ERROR;
		case 'f':
			flags |= BF_FORCE;
			break;
		case 'y':
			flags |= BF_YES;
			break;
		case 'l':
			flags |= BF_SHOW_PLUG;
			break;
		case 'p':
			flags |= BF_SHOW_PARM;
			break;
		case 's':
			flags |= BF_LOST;
			break;
		case 'd':
			flags |= BF_DISCARD;
			break;
		case 'm':
			flags |= BF_MIRRORS;
			break;
		case 'o':
			aal_strncat(override, optarg,
				    aal_strlen(optarg));
			
			aal_strncat(override, ",", 1);
			break;
		case 'b':
			/* Parsing blocksize */
			if ((hint.blksize = misc_str2long(optarg, 10)) == INVAL_DIG) {
				aal_error("Invalid blocksize (%s).", optarg);
				return USER_ERROR;
			}
			
			if (!aal_pow2(hint.blksize)) {
				aal_error("Invalid blocksize (%u). It must power "
					  "of two.", hint.blksize);
				return USER_ERROR;	
			}
			break;
		case 'U':
			/* Parsing passed by user uuid */
			if (aal_strlen(optarg) != 36) {
				aal_error("Invalid uuid was specified (%s).",
					   optarg);
				return USER_ERROR;
			}
#if defined(HAVE_LIBUUID) && defined(HAVE_UUID_UUID_H)
			{
			  if (uuid_parse(optarg,
					 (unsigned char *)hint.volume_uuid) < 0) {
					aal_error("Invalid uuid was "
						  "specified (%s).",
						  optarg);
					return USER_ERROR;
				}
			}
#endif
			break;
		case 'L':
			aal_strncpy(hint.label, optarg,
				    sizeof(hint.label));
			break;
		}
	}

	if (!(flags & BF_YES))
		misc_print_banner(argv[0]);

	/* Initializing libreiser4 (getting plugins, checking them on validness,
	   etc). */
	if (libreiser4_init()) {
		aal_error("Can't initialize libreiser4.");
		goto error;
	}

	/* Overriding default params by passed values. This should be done after
	   libreiser4 is initialized. */
	if (aal_strlen(override) > 0) {
		override[aal_strlen(override) - 1] = '\0';

		if (!(flags & BF_YES)) {
			aal_mess("Overriding the plugin profile by \"%s\".",
				 override);
		}

		if (misc_profile_override(override))
			goto error_free_libreiser4;
	}

	if (flags & BF_SHOW_PARM)
		misc_profile_print();
	
	if (flags & BF_SHOW_PLUG)
		misc_plugins_print();

	if (optind >= argc)
		goto error_free_libreiser4;
	
	if (optind >= argc) {
		mkfs_print_usage(argv[0]);
		return USER_ERROR;
	}
    

#ifdef HAVE_SYSCONF
	/* Guessing block size by getting page size */
	if (!hint.blksize) {
		hint.blksize = sysconf(_SC_PAGESIZE);
	} else {
		if (!(flags & BF_FORCE)) {
			if (hint.blksize != (uint32_t)sysconf(_SC_PAGESIZE)) {
				aal_warn("Block size (%u) and page size "
					 "(%ld) mismatch is detected. "
					 "Reiser4 does not support block "
					 "sizes different than page size "
					 "yet. Use -f to force over.",
					 hint.blksize, sysconf(_SC_PAGESIZE));
				goto error_free_libreiser4;
			}
		}
	}
	
	if (!(flags & BF_YES)) {
		aal_mess("Block size %u will be used.", hint.blksize);
	}
#else
	if (!hint.blksize) {
		aal_warn("Can't guess page size. Default "
			 "block size (4096) will be used, "
			 "or use -b option instead.");
		hint.blksize = 4096;
	}
#endif
	
	if (hint.blksize > REISER4_MAX_BLKSIZE) {
		aal_error("Invalid blocksize (%u). It must not be greater then "
			  "%u.", hint.blksize, REISER4_MAX_BLKSIZE);
		return USER_ERROR;	
	}
	
#ifdef HAVE_UNAME
	/* Guessing system type */
	if (uname(&sysinfo) == -1) {
		aal_warn("Can't guess system type.");
		goto error_free_libreiser4;
	}

	if (!(flags & BF_YES)) {
		aal_mess("%s %s is detected.", sysinfo.sysname,
			 sysinfo.release);
	}
#endif
	/*
	 * Building list of devices the filesystem will be created on
	 */
	for (; optind < argc; optind++) {
		if (stat(argv[optind], &st) == -1) {
			aal_error("Can't stat %s. %s.", argv[optind],
				  strerror(errno));
			goto error_free_libreiser4;

			if (misc_size2long(argv[optind]) != INVAL_DIG &&
			    hint.blocks != 0)
			{
				aal_error("Filesystem length is already "
					  "set to %llu.", hint.blocks);
				continue;
			}
			/*
			 * Checking device name for validness
			 */
			hint.blocks = misc_size2long(argv[optind]);

			if (hint.blocks != INVAL_DIG) {
				/* Converting into fs blocksize blocks */
				hint.blocks /= (hint.blksize / 1024);
				/* Just to know that blocks was given. 0 
				   means nothing was specified by user. */
				if (!hint.blocks) hint.blocks = 1;
			} else {
				aal_error("%s is not a valid size nor an "
					  "existent file.", argv[optind]);
				goto error_free_libreiser4;
			}
		} else {
			devices = aal_list_append(devices, argv[optind]);
			dev_cnt ++;
		}
	}
	/*
	 * Set number of subvolumes and mirrors
	 */
	if (flags & BF_MIRRORS)
		hint.num_subvols = dev_cnt;
	else
		hint.num_subvols = 1;
	hint.num_mirrors = hint.num_subvols - 1;

	dev_cnt = 0;
	if (!(flags & BF_YES) && aal_list_len(devices)) {
		if (!(gauge = aal_gauge_create(aux_gauge_handlers[GT_PROGRESS],
					       NULL, NULL, 0, NULL)))
			goto error_free_libreiser4;
	}
  	/*
	 * Loop through all the accumulated devices
	 */
	aal_list_foreach_forward(devices, walk) {
    
		host_dev = (char *)walk->data;
    
		if (stat(host_dev, &st) == -1) {
			aal_error("Can't stat %s. %s.", host_dev,
				  strerror(errno));
			goto error_free_libreiser4;
		}
		/*
		 * Check if passed device is a block device. If so, we check
		 * also is it whole drive or just a partition. If the device is
		 * not a block device, then we emmit exception and propose user
		 * to use -f flag to force
		 */
		if (!S_ISBLK(st.st_mode)) {
			if (!(flags & BF_FORCE)) {
				aal_error("Device %s is not block "
					  "device. Use -f to force "
					  "over.", host_dev);
				goto error_free_libreiser4;
			}
		} else {
			if (((IDE_DISK_MAJOR(MAJOR(st.st_rdev)) &&
			      MINOR(st.st_rdev) % 64 == 0) ||
			     (SCSI_BLK_MAJOR(MAJOR(st.st_rdev)) &&
			      MINOR(st.st_rdev) % 16 == 0)) &&
			    !(flags & BF_FORCE))
			{
				aal_error("Device %s is an entire "
					  "harddrive, not just one "
					  "partition.", host_dev);
				goto error_free_libreiser4;
			}
		}
   		/*
		 * Check if passed partition is mounted
		 */
		if (misc_dev_mounted(host_dev) > 0 && !(flags & BF_FORCE)) {
			aal_error("Device %s is mounted at the moment. "
				  "Use -f to force over.", host_dev);
			goto error_free_libreiser4;
		}
		/*
		 * Generate uuid if it was not specified and
		 * if libuuid is in use
		 */
#if defined(HAVE_LIBUUID) && defined(HAVE_UUID_UUID_H)
		if (uuid_is_null((unsigned char *)hint.volume_uuid)) {
			uuid_generate((unsigned char *)hint.volume_uuid);
		}
		if (uuid_is_null((unsigned char *)hint.subvol_uuid)) {
			uuid_generate((unsigned char *)hint.subvol_uuid);
		}
		if (!(flags & BF_YES)) {
			char uuid[256];

			uuid_unparse((unsigned char *)hint.volume_uuid, uuid);
			aal_mess("Volume uuid %s will be used.", uuid);
		}
#else
		if (flags & BF_MIRRORS) {
			aal_error("uuid is required to create logical volumes");
			goto error_free_libreiser4;
		}
#endif
		/*
		 * Initialize mkfs_id for fsck needs
		 */
		if (hint.mkfs_id == 0) {
			srandom(time(0) + dev_cnt);
			hint.mkfs_id = random();
		}
		if (flags & BF_MIRRORS)
			hint.subvol_id = dev_cnt;

		if (!(device = aal_device_open(&file_ops, host_dev, 
					       512, O_RDWR))) 
		{
			aal_error("Can't open %s. %s.",
				  host_dev, strerror(errno));
			goto error_free_libreiser4;
		}
    		/*
		 * Convert device length into fs blocksize blocks
		 */
		dev_len = reiser4_format_len(device, hint.blksize);

		if (flags & BF_MIRRORS &&
		    hint.blocks != 0 && hint.blocks != dev_len) {
			aal_error("Mirror can be created only on a device "
				  "of the same length as the original has.");
			goto error_free_device;
		}
		if (hint.blocks > dev_len) {
			aal_error("Filesystem wouldn't fit into device "
				  "%llu blocks long, %llu blocks required.",
				  dev_len, hint.blocks);
			goto error_free_device;
		}
		if (hint.blocks == 0)
			hint.blocks = dev_len;
		/*
		 * Check for non-intercative mode
		 */
		if (!(flags & BF_YES)) {
			if (aal_yesno("Reiser4 is going to be created on %s.",
				      host_dev) == EXCEPTION_OPT_NO)
			{
				goto error_free_device;
			}
		}
		if (flags & BF_DISCARD) {
			if (gauge) {
				aal_gauge_rename(gauge, "Discarding %s ... ",
						 host_dev);
				aal_gauge_touch(gauge);
			}

			if ((aal_device_discard(device, 0, dev_len) != 0)) {
				aal_error("Failed to discard %s (%s).",
					  device->name, device->error);
				/* discard is optional, don't fail the mkfs */
			}

			if (gauge) {
				aal_gauge_done(gauge);
			}
		}

		if (gauge) {
			aal_gauge_rename(gauge, "Creating reiser4 on %s ... ",
					 host_dev);
			aal_gauge_touch(gauge);
		}
		/*
		 * Create a filesystem on the device
		 */
		if (!(fs = reiser4_fs_create(device, &hint))) {
			aal_error("Can't create filesystem on %s.", 
				  device->name);
			goto error_free_device;
		}

		/* Creating journal */
		if (!(fs->journal = reiser4_journal_create(fs, device)))
			goto error_free_fs;

		/* Creating root directory */
		if (!(fs->root = reiser4_root_create(fs))) {
			aal_error("Can't create filesystem "
				  "root directory.");
			goto error_free_fs;
		}

		/* Linking root to itself */
		if (reiser4_object_link(fs->root, fs->root, NULL)) {
			aal_error("Can't link root directory "
				  "to itself.");
			goto error_free_fs;
		}
	
		if (reiser4_pset_tree(fs->tree, 1)) {
			aal_error("Can't initialize the fs-global "
				  "object plugin set.");
			goto error_free_fs;
		}

		/* Backup the fs metadata. */
		if (!(fs->backup = reiser4_backup_create(fs))) {
			aal_error("Can't create the fs metadata backup.");
			goto error_free_fs;
		}

		/* Creating lost+found directory */
		if (flags & BF_LOST) {
			reiser4_object_t *object;
	    
			if (!(object = reiser4_dir_create(fs->root,
							  "lost+found")))
			{
				aal_error("Can't create \"/lost+found\" "
					  "directory.");
				goto error_free_root;
			}
	    
			reiser4_object_close(object);
		}
	
		if (gauge)
			aal_gauge_done(gauge);

		if (!(flags & BF_MIRRORS)) {
			/*
			 * The following parameters are independent,
			 * so we zero them in order to force mkfs to
			 * generate different ones for next device
			 * form built device list
			 */
			aal_memset(hint.volume_uuid, 0, sizeof(hint.volume_uuid));
			aal_memset(hint.label, 0, sizeof(hint.label));
			hint.mkfs_id = 0;
			hint.blocks = 0;
		}
		aal_memset(hint.subvol_uuid, 0, sizeof(hint.subvol_uuid));
		/*
		 * Free the root directory & fs
		 */
		reiser4_object_close(fs->root);
		reiser4_fs_close(fs);
		/*
		 * Synchronizing device.
		 * If device we are using is a file device (libaal/file.c),
		 * then function fsync will be called
		 */
		if (aal_device_sync(device)) {
			aal_error("Can't synchronize device %s.", 
				  device->name);
			goto error_free_device;
		}

		aal_device_close(device);
		dev_cnt ++;
	}
    
	/* Freeing the all used objects */
	if (gauge)
		aal_gauge_free(gauge);
	
	aal_list_free(devices, NULL, NULL);

	/* Deinitializing libreiser4. At the moment only plugins are unloading
	   during this. */
	libreiser4_fini();
    
	return NO_ERROR;

 error_free_root:
	reiser4_object_close(fs->root);
 error_free_fs:
	reiser4_fs_close(fs);
 error_free_device:
	aal_device_close(device);
 error_free_libreiser4:
	libreiser4_fini();
 error:
	return OPER_ERROR;
}
